[AI] CSB remote skill (#179)
open-s4c/csb: .agents/skills/csb-remote/SKILL.md

---
name: csb-remote
description: "Use when CSB work is executed on one or more remote benchmark machines over SSH while the current host acts as the AI/controller host: preparing remote CSB checkouts and dependencies, installing benchmark/monitor tooling, setting permissions, syncing host skills/templates/configs, running CSB usage/analysis/refine/dev commands remotely, handling reconnects and cleanup, coordinating multiple remote machines, and copying results/reports/configs/monitor artifacts back into remote-specific result directories on the host."
---

# CSB Remote

> 

Reviewer: I wonder if we can also let the skill help the user configure ssh if it fails.

Author: yes, thats in

Use this skill whenever a CSB benchmark, refinement, analysis rerun, monitor capture, source checkout, build, or validation is performed on an SSH-accessible remote machine instead of the controller host running Codex.

## Roles

- **Controller host**: the machine running the AI agent. It owns the Codex skills, local instructions, final copied reports, and cross-remote organization.
- **Remote benchmark host**: the machine that executes CSB commands, benchmarks, monitors, package installs, source checkouts, kernel-source correlation, builds, and cleanup.
- Treat the remote as the source of truth for benchmark outputs, system/kernel facts, monitor availability, and tested source trees. The controller can generate or edit skill-guided scripts/configs, but every runtime check and benchmark command must be executed remotely.

## Skill Composition

Use this skill with the relevant CSB skill:

> - `csb`: choose/adapt configs and run campaigns remotely.

Reviewer 2: Does it still adapt the configs?

Author: the csb skill does not, just csb-refine

- `csb-analysis`: analyze remote result artifacts and perform source/perf inspection against the remote's tested kernel/source tree.
- `csb-refine`: perform every sweep, cleanup, monitor adaptation, diagnostic run, report generation, and patch/object-build test remotely.

> - `csb-dev`: reproduce framework bugs, run tests, and validate monitor/runner changes remotely when the behavior depends on remote-only permissions, hardware, kernel, containers, perf, bpftrace, or topology.

Reviewer: ```suggestion
```

I don't think it should carry out development on remote.

> When instructions in another CSB skill say "host", "machine", "testing machine", "local command", or "prepare the environment", interpret that as the remote benchmark host for remote work. Keep only final aggregation, copied artifacts, and skill files on the controller unless explicitly asked to edit the remote checkout.

Reviewer 2: What is the context for "local command" here? I'm a bit struggling to see the example where "local command" has to be executed remotely.

## Remote Inventory

> For each remote, record a short inventory before running CSB:

Reviewer 2: ```suggestion
For each remote host, record a short inventory before running CSB:
```
I believe it'd be better to not sound like "git remotes". Can we switch away from "remote" to "remote host"/etc.?

```bash
ssh <remote> 'hostname; pwd; uname -a; cat /etc/os-release 2>/dev/null || true; nproc; lscpu | sed -n "1,80p"'
ssh <remote> 'command -v git python3 sudo perf bpftrace docker runc youki mpstat iostat || true'
ssh <remote> 'test -d /path/to/csb && git -C /path/to/csb status --short || true'
```

Keep this mapping in notes or reports:

- remote name and SSH target;
- remote CSB checkout path;
- remote result group/root;
- remote kernel/source tree path, usually `deps/linux`;
- runtime/container tools and monitor permissions;
- any node-specific cleanup/runtime quirks.

> Multiple remotes may be used simultaneously. Prefix tmux sessions, result groups, temp configs, and copied directories with the remote name so artifacts do not collide.

Reviewer 2: Should we add something about not touching existing user's tmux sessions? I see later on it uses a specific format for the tmux session name though...

Author: yes, a csb-remote namespace

## Remote Setup

The remote must have a complete CSB setup with all dependencies needed by the selected CSB skills. Execute the checks and setup on the remote:

```bash
ssh <remote> 'cd /path/to/csb && scripts/prepare.sh'
ssh <remote> 'cd /path/to/csb && cmake -S. -Bbuild && cmake --build build -j"$(nproc)"'

> ssh <remote> 'cd /path/to/csb && python3 -m json.tool config/<file>.json >/dev/null'

Reviewer: ```suggestion
```

- The bm-runner build what it needs, this is an overkill to add here.
- Also the JSON file to be run should be checked on controller before the copying not on the remote.

```

Install missing software on the remote, not the controller, when needed for the requested benchmark or monitors. Typical packages/tools include Python build dependencies, CMake, Docker or the selected OCI runtime, sysstat (`mpstat`/`iostat`), `perf`, `bpftrace`, benchmark-specific `*-dev` packages, and kernel build/source dependencies.

For perf, tracefs, bpftrace, Arm SPE, containers, cgroups, NICs, and kernel-source setup, run the permission and availability commands from `csb`, `csb-analysis`, and `csb-refine` on the remote. Do not infer remote capability from the controller.

## Source Trees

Kernel source used for source correlation and object-build tests belongs on the remote unless the user explicitly asks for host-only analysis. Use one remote `deps/linux` tree:

- add or update a Torvalds remote for upstream comparison;
- add or update the tested distro/vendor/kernel remote;
- check out the branch/tag/commit closest to the running remote kernel;
- record remotes, checked-out commit, dirty state, and comparison ref in reports.

> If host skills provide bpftrace templates, helper scripts, or report generators that are not present remotely, copy those files to a temporary remote path or into the remote CSB checkout before running. Treat controller copies as templates; executed scripts and generated programs should live in the remote run artifacts and be copied back afterward.

Reviewer: Why on the remote? searching for relevant patches can be on controller right? after all results are copied?

## Running Remotely

Run CSB through SSH from the remote CSB checkout:

```bash
ssh <remote> 'cd /path/to/csb && CSB_RESULTS_GROUP=<remote>_<group> scripts/run-single.sh config/<file>.json'
```

> For long runs, use `tmux` or `systemd-run` on the remote and reconnect until completion:

Reviewer: ```suggestion
Always use `tmux` or `systemd-run` on the remote and reconnect until completion:
```

```bash
ssh <remote> 'cd /path/to/csb && tmux new-session -d -s csb_<remote>_<name> "CSB_RESULTS_GROUP=<group> scripts/run-single.sh config/<file>.json 2>&1 | tee results/<group>.log"'
ssh <remote> 'tmux capture-pane -pt csb_<remote>_<name> -S -200'
```

If the SSH connection drops, reconnect and inspect the remote session/processes. Do not restart a sweep until you know whether the prior sweep is still running, completed, or failed and cleanup is needed.

## Cleanup Between Sweeps

Before every new remote CSB sweep, perform the cleanup required by `csb-refine` on that remote. Clean only artifacts belonging to the prior CSB sweep:

- stale benchmark/application processes;
- stale Docker/runc/youki/runtime state with the benchmark prefix;
- stale monitor processes such as `perf`, `perf lock`, `bpftrace`, `mpstat`, `iostat`;
- CSB start barriers such as `build/bench/start`;
- temporary monitor files, oversized partial perf data, and scratch dirs not meant to be kept.

> Re-check after cleanup with process/runtime-specific commands. Record the cleanup summary in the remote run report.

Reviewer: but that's already mentioned partially in the refine/other skills. I would rather not repeat it here.

## Copy-Back Layout

At the end of each remote task, copy all final results, reports, configs, generated bpftrace programs, helper scripts, and patch artifacts back to the controller under a remote-specific directory. Prefer this layout:

```text
results/remote/<remote>/<task-or-group>/
  results/          # copied remote results/<group> or selected run dirs
  reports/          # summary/detailed/cross-run reports
  configs/          # configs used for the remote runs
  monitors/         # generated bpftrace programs, perf scripts, helper probes
  patches/          # RFC patches and patch support docs, if any
  logs/             # driver logs, tmux captures, install/setup notes
  inventory/        # uname, lscpu, os-release, tool versions, git refs

> ```

Reviewer: I think it should only need the results and all the rest is either already there or will be generated by AI on controller.

Author: hm, i am not sure about that, it might end up mixing stuff. let it generate all on the remote and also have a copy of it on the controller at the end so the state is not split between different machines

- For multiple remote hosts, prefix tmux sessions, result groups, temp configs, and copied directories with `csb-remote_<remote>_`.
- Do not kill, rename, or reuse unrelated user tmux sessions.
- If SSH drops, reconnect and inspect remote sessions/processes before restarting or cleaning up.

## Minimal Inventory

Record this before running CSB and include it in reports:

```bash
ssh <remote> 'hostname; pwd; uname -a; cat /etc/os-release 2>/dev/null || true; nproc; lscpu | sed -n "1,80p"'
ssh <remote> 'command -v git python3 sudo perf bpftrace docker runc youki mpstat iostat || true'
ssh <remote> 'test -d /path/to/csb && git -C /path/to/csb status --short || true'
```

Also track: SSH target, remote CSB checkout path, result group/root, remote running-kernel identity, controller kernel-source path/ref used for that host, runtime and monitor permissions, and any node-specific cleanup quirks.

## Execution Pattern

Use SSH from the controller and execute from the remote CSB checkout:

```bash
ssh <remote> 'cd /path/to/csb && CSB_RESULTS_GROUP=<remote>_<group> scripts/run-single.sh config/<file>.json'
```

> For long runs, prefer a remote `tmux` or `systemd-run` session:

Author: ```suggestion
Always use a remote `tmux` or `systemd-run` session:
```
